feature/parallelise-postcode-searches

AdamJSc opened the pull request:
The purpose of this change is to shift the focus from Jobs to Tasks in terms of maintaining state, configuration, execution and monitoring.

This enables multiple Task Payloads to be associated with a single Job. The Job's Task function can then be run multiple times, taking in a different Payload each time.

Each Payload now becomes the single unit that we wish to maintain state for independently, rather than the Job.

Each Task-Payload combination can now be executed concurrently, in parallel to the other Task-Payload combinations within the same Job, as well as any other Jobs.

Other changes include:
- Make delay/interval between Task runs configurable on a per-Task basis. Also randomise this slightly, between an upper/lower bound that is determined by a numeric global offset either side of the task interval (e.g. +/- 7 seconds).
- Configure Task bypass as a timestamp rather than a boolean. This enables a Task to be bypassed for a specific duration of time, rather than indefinitely until a new Task State file takes over (i.e. the next day)

AdamJSc committed:
Introduce TaskPayload

AdamJSc committed:
Parse TaskPayloads in main bootstrap

AdamJSc committed:
Orient the processing of jobs around Tasks and TaskPayloads.

Enables control of state, intervals and execution to occur
at child Task level rather than parent Job level.

AdamJSc committed:
Clean up data folder debris

AdamJSc committed:
Randomise interval between two consecutive task runs.

Define an offset that provides an upper/lower limit to randomise by,
either side of the task's base interval.

AdamJSc committed:
Ensure that processing first `TaskPayload` doesn't block subsequent payloads.

Run the processing of each payload within its own goroutine, then
block indefinitely immediately after the last iteration.

AdamJSc committed:
Update README and example data

AdamJSc committed:
Amend task-related variable names

AdamJSc committed:
Move Asda API URL from struct field to a constant

AdamJSc committed:
Fix task interval randomisation and taskstate filename for ordering by date

AdamJSc committed:
Replace taskstate `Bypass` bool field with `BypassUntil` timestamp.

Also determine whether to bypass task within `runTask` bootstrap instead.

This removes the onus of needing to perform this check within individual task functions,
our state is now loaded in advance and can be checked prior to passing it to the function on each invocation.

AdamJSc committed:
Amend default bypass to represent minutes not seconds + update README

ladydascalie reviewed (approved):
I think this is very good work :)

AdamJSc committed:
Amend date range look-ahead when checking for available slots.

AdamJSc committed:
When filtering a `DeliveryManifest` by availability, only append a filtered `DailySchedule` if it has at least one slot.

This prevents a filtered manifest from comprising `DailySchedules` that have no available slots.
